app/models/robot_directory.rb

```diff
@@ -1,59 +1,101 @@
-      directory['robots'] << { 'serial_number' => directory['total'] * rand(100),
-                               'name' => robot[:name],
-                               'city' => robot[:city],
-                               'state' => robot[:state],
-                               'avatar' => robot[:avatar],
-                               'birthdate' => robot[:birth_date],
-                               'date_hired' => robot[:date_hired],
-                               'department' => robot[:department] }
-    end
+    directory.execute("INSERT INTO robots
+      (name, city, state, avatar, birthdate, date_hired, department)
+      VALUES
+      (?, ?, ?, ?, ?, ?, ?);", robot[:name], robot[:city], robot[:state],
+      'robohash', robot[:birthdate], robot[:date_hired], robot[:department])
+
+    serial = directory.execute("SELECT serial_number
+                                      FROM robots
+                                      WHERE name=?;", robot[:name]).flatten.last['serial_number']
+    directory.execute("UPDATE robots
+                      SET avatar='https://robohash.org/#{serial}'
+                      WHERE name=?;", robot[:name])
   end
 
   def all
-    directory.transaction do
-      robots.map { |robot| Robot.new(robot) }
-    end
+    raw_robots.map { |robot| Robot.new(robot) }
   end
 
-  def robots
-    directory.transaction do
-      database['robots'] || []
-    end
+  def raw_robots
+    directory.execute("SELECT * FROM robots;")
   end
 
-  def robot(serial_number)
-    robots.find { |robot| robot['serial_number'] == serial_number }
+  def raw_robot(serial_number)
+    directory.execute("SELECT * FROM robots
+                       WHERE serial_number=?;", serial_number).first
+  end
+
+  def find(serial_number)
+    Robot.new(raw_robot(serial_number))
   end
 
   def update(serial_number, robot)
-    directory.transaction do
-      target_robot = directory['robots'].find { |robot| robot['serial_number'] == serial_number }
-      target_robot['name']       = robot[:name]
-      target_robot['city']       = robot[:city]
-      target_robot['state']      = robot[:state]
-      target_robot['avatar']     = robot[:avatar]
-      target_robot['birthdate']  = robot[:birthdate]
-      target_robot['date_hired'] = robot[:date_hired]
-      target_robot['department'] = robot[:department]
-    end
+    directory.execute("UPDATE robots
+                       SET name=?, city=?, state=?, birthdate=?, date_hired=?,
+                       department=? WHERE serial_number=?;", robot[:name],
+                       robot[:city], robot[:state], robot[:birthdate],
+                       robot[:date_hired], robot[:department], serial_number)
   end
 
   def delete(serial_number)
-    directory.transaction do
-      directory['robots'].delete_if { |robot| robot['serial_number'] == serial_number }
-    end
+    directory.execute("DELETE FROM robots WHERE serial_number=?;", serial_number)
+  end
+
+  def average_age
```

Comment:
if we want to be really single responsibility, we have two things going on - finding all the ages, then finding the average. i suggest even pulling these separate behaviors into separate methods